tempesta_fw/str.h

```diff
@@ -98,40 +94,45 @@
  * @flags	- 3 most significant bytes for number of chunks of compound
  * 		  string and the least significant byte for flags;
  */
-typedef struct {
-	void		*ptr;
+typedef struct TfwStr {
 	struct sk_buff	*skb;
 	unsigned long	len;
 	unsigned char	eolen;
-	unsigned int	flags;
+
+	unsigned int	chunknum : 24;
+	unsigned int	flags : 8;
+	union {
+	char *data;
+	struct TfwStr *chunks;
+};
 } TfwStr;
 
-#define DEFINE_TFW_STR(name, val) TfwStr name = { (val), NULL,		\
-						  sizeof(val) - 1, 0 }
-#define TFW_STR_FROM(s)         ((TfwStr){(char*)s, NULL, strlen(s)})
+#define DEFINE_TFW_STR(name, val) TfwStr name = { .data = (val), .skb = NULL\
+, .len = sizeof(val) - 1, .chunknum = 0, .flags = 0 }
+#define TFW_STR_FROM(s)         ((TfwStr){ .data = (char *)s, .skb = NULL,\
+  .len = sizeof(s) - 1})
 
 /* Use this with "%.*s" in printing calls. */
-#define PR_TFW_STR(s)		(int)min(20UL, (s)->len), (char *)(s)->ptr
+#define PR_TFW_STR(s)		(int)min(20UL, (s)->len), (s)->data
 
 /* Numner of chunks in @s. */
-#define TFW_STR_CHUNKN(s)	((s)->flags >> TFW_STR_CN_SHIFT)
+#define TFW_STR_CHUNKN(s)	((s)->chunknum)
 #define TFW_STR_CHUNKN_LIM(s)	((s)->flags >= __TFW_STR_CN_MAX)
-#define TFW_STR_CHUNKN_ADD(s, n) ((s)->flags += ((n) << TFW_STR_CN_SHIFT))
-#define TFW_STR_CHUNKN_SUB(s, n) ((s)->flags -= ((n) << TFW_STR_CN_SHIFT))
-#define __TFW_STR_CHUNKN_SET(s, n) ((s)->flags = ((s)->flags & TFW_STR_FMASK) \
-						  | ((n) << TFW_STR_CN_SHIFT))
+#define TFW_STR_CHUNKN_ADD(s, n) ((s)->chunknum += (n))
+#define TFW_STR_CHUNKN_SUB(s, n) ((s)->chunknum -= (n))
+#define __TFW_STR_CHUNKN_SET(s, n) ((s)->chunknum = (n))
 /* Compound string contains at least 2 chunks. */
 #define TFW_STR_CHUNKN_INIT(s)	__TFW_STR_CHUNKN_SET(s, 2)
 
 #define TFW_STR_INIT(s)		memset(s, 0, sizeof(TfwStr))
 
 #define TFW_STR_EMPTY(s)	(!((s)->flags | (s)->len))
-#define TFW_STR_PLAIN(s)	(!((s)->flags & __TFW_STR_COMPOUND))
+#define TFW_STR_PLAIN(s)	((s)->chunknum > 0)
```

Comment:
First of all, `chunknum` is unsigned and can never be less than zero. Secondly, a `TfwStr{}` string is plain when is **does not** have chunks. So, the corresponding statement in C is just:
```C
(!(s)->chunknum)
```